shell/platform/darwin/macos/framework/Source/FlutterEngine.mm

```diff
@@ -834,13 +834,14 @@ - (void)updateDisplayConfig {
     CGDirectDisplayID displayID =
         static_cast<CGDirectDisplayID>([screen.deviceDescription[@"NSScreenNumber"] integerValue]);
 
+    double devicePixelRatio = screen.backingScaleFactor;
     FlutterEngineDisplay display;
     display.struct_size = sizeof(display);
     display.display_id = displayID;
     display.single_display = false;
-    display.width = static_cast<size_t>(screen.frame.size.width);
-    display.height = static_cast<size_t>(screen.frame.size.height);
-    display.device_pixel_ratio = screen.backingScaleFactor;
+    display.width = static_cast<size_t>(screen.frame.size.width) * devicePixelRatio;
+    display.height = static_cast<size_t>(screen.frame.size.height) * devicePixelRatio;
+    display.device_pixel_ratio = devicePixelRatio;
```

Comment:
For testing this, I think we could probably extract the fetching of screens out as a method in the engine_internal.h header. In the test, we use a subclass that overrides the method providing the screens, call updateDisplayConfig, and verify we got a NotifyDisplayUpdate call to the embedder API with the expected size/resolutions.

Reply:
Hm, the test fails on line 1183 indicating that the embedder's `NotifyDisplayUpdate` is never called. I traced it down to line 832 above: _engine is null at this point and we hit the early return. Any idea what my test is missing to get FlutterEngine into the right state where _engine is not null?

Reply:
Oh, I think I am missing a call to `[engine runWithEntrypoint:@"main"]`. Let me try that.

Reply:
Awesome looks great!